pelican_bib/bib.py

```
@@ -160,6 +201,159 @@ def add_publications(generator):
+        'sorting_style': directives.unchanged,
+        'abbreviate_names': boolean,
+        'name_style': directives.unchanged
+    }
+
+    def run(self):
+        refs_files = self.arguments
+        refs_string = '\n'.join(self.content)
+        template_name = self.generator.settings.get(
+            'PUBLICATIONS_DEFAULT_TEMPLATE', 'bibliography'
+        )
+        template_options = {}
+        classes = ['bibliography']
+        filter_tag = None
+        pybtex_style_args = {}
+
+        # fetch arguments
+        if refs_files:
+            classes += (basename(file) for file in refs_files)
+        if 'template' in self.options:
+            template_name = self.options['template']
+        if 'options' in self.options:
+            template_options = literal_eval(self.options['options'])
+        if 'class' in self.options:
+            classes = self.options['class']
+        if 'filter_tag' in self.options:
+            filter_tag = self.options['filter_tag']
+        if 'pybtex_style_args' in self.options:
+            pybtex_style_args.update(literal_eval(self.options['pybtex_style_args']))
+        if 'sorting_style' in self.options:
+            pybtex_style_args['sorting_style'] = self.options[
+                'sorting_style'
+            ]
+        if 'abbreviate_names' in self.options:
+            pybtex_style_args['abbreviate_names'] = self.options['abbreviate_names']
+        if 'name_style' in self.options:
+            pybtex_style_args['name_style'] = self.options['name_style']
+
+        # BibTeX input is a bib file or the directives content
+        if not self.content and not refs_files:
+            raise self.error('No BibTeX input source given.')
+
+        refs_files = (self.get_pelican_path(file) for file in refs_files)
+
+        # create a copy of generator context (don't mess up original
+        # context)
+        original_context = self.generator.context
+        self.generator.context = self.generator.context.copy()
```

Why do you keep the copy? If there is an error in rendering, I would expect the plugin to just fail and to not create any output and let me fix whatever I have to fix.

---

+1 for not failing silently

---

The original context is created once per pelican build, the directive's context is created once per usage of the directive.
So creating a copy is importing for two reasons:

1. Using a directive would make it impossible to use pelican_bib as it has been used so far (`:template: publications`), because most likely the context will be overwritten by the directive before it could be used by the template.

2. If the context would _not_ be restored, an error would potentially occur in all subsequently processed directives. So it would fail the opposite of "silently" - there would be a lot of error messages messing up the output.